### ISSUE TITLE
proof and theorem chunk engine does not work anymore

### ISSUE THREAD

**Issue by cderv** [rstudio/bookdown#1178]
I am currently writing a bookdown book which already has >400 pages. It has numerous theorem and proof chunks, and since a couple of days these do not work any more (Rstudio 1.4.1717, latest bookdown from Github, OS X Big Sur 11.5 beta). For some reason numerous \iffalse ... \fi are inserted, even in the gitbook output format. See example .Rmd file (disguised as .txt).

[index.txt](https://github.com/rstudio/bookdown/files/6595318/index.txt)

````markdown
---
title: "KoKo"
author: "Jan de Leeuw"
date: '`r paste("Started June 02, 2021. Last update", format(Sys.Date(),"%B %d, %Y"))`'
site: "bookdown::bookdown_site"
output: bookdown::pdf_book
---

# Ozo

```{theorem, atheorem, name = "A Theorem"}
This is a theorem.
```

```{proof}
This is a proof of theorem \@ref(thm:atheorem)
```
````

_Originally posted by @deleeuw in https://github.com/rstudio/bookdown/issues/1166#issuecomment-854305818_

**Comment by cderv** [comment]
There is an issue with how block2 engine is working which cause the proof and theorem chunk to fail. I need to find where it / since when it fails.

**Comment by deleeuw** [comment]
I realize this does not solve the problem, but I transformed it away in the book by using the 

:::  {.foo}  bar ::: 

syntax throughout.

**Comment by cderv** [comment]
Thanks for the feedback. Glad the new syntax helped you ! 

I looked further and this could be an issue caused by Pandoc. With last version (2.14.0.1) I get this error 

````
! Extra \fi.
l.108 ...65-32-84-104-101-111-114-101-109-93-\}\fi
                                                  {} 

Error: LaTeX failed to compile test.tex. See https://yihui.org/tinytex/r/#debugging for debugging tips. See test.log for more info.
````

But I can render it correctly by juts switching to an older version like Pandoc 2.11.4 shipped with RStudio IDE. 

So this is a probably a regression due to Pandoc update (but I don't understand why we did not catched that). 

Can you share with us the pandoc version your are using please ? 

by running

````r
rmarkdown::pandoc_version()
````

Thanks

**Comment by cderv** [comment]
@yihui I did look further into Pandoc changes

This is what knitr will generate in intermediate `.md` for the `theorem` engine

````markdown
\BeginKnitrBlock{theorem}\iffalse{-91-65-32-84-104-101-111-114-101-109-93-}\fi{}
<span class="theorem" id="thm:atheorem"><strong>(\#thm:atheorem)  \iffalse (A Theorem) \fi{} </strong></span>This is a theorem.
\EndKnitrBlock{theorem}
````

This has not changed but pandoc will handle that differently in newer version
With Pandoc 2.11 for example

````markdown
❯ C:\Users\chris\Logiciel\pandoc\pandoc-2.11\pandoc.exe -t latex
\BeginKnitrBlock{theorem}\iffalse{-91-65-32-84-104-101-111-114-101-109-93-}\fi{}
<span class="theorem" id="thm:atheorem"><strong>(\#thm:atheorem)  \iffalse (A Theorem) \fi{} </strong></span>This is a theorem.
\EndKnitrBlock{theorem}
^Z
\BeginKnitrBlock{theorem}\iffalse{-91-65-32-84-104-101-111-114-101-109-93-}\fi{}

\protect\hypertarget{thm:atheorem}{}{(\#thm:atheorem) \iffalse (A
Theorem) \fi{} }This is a theorem. \EndKnitrBlock{theorem}
````
It is as expected for our `restore_block2()` function

With last Pandoc this is now broken 

````powershell
❯ pandoc -t latex
\BeginKnitrBlock{theorem}\iffalse{-91-65-32-84-104-101-111-114-101-109-93-}\fi{}
<span class="theorem" id="thm:atheorem"><strong>(\#thm:atheorem)  \iffalse (A Theorem) \fi{} </strong></span>This is a theorem.
\EndKnitrBlock{theorem}
^Z
\BeginKnitrBlock{theorem}{}

ffalse\{-91-65-32-84-104-101-111-114-101-109-93-\}\fi{}
\protect\hypertarget{thm:atheorem}{}{(\#thm:atheorem)
\textbackslash iffalse (A Theorem) \fi{} }This is a theorem.
\EndKnitrBlock{theorem}
````

Here is the diff for easier look at the changes

````diff
diff --git a/test.md b/test.md
index 681872d5..a7d40939 100644
--- a/test.md
+++ b/test.md
@@ -11,7 +11,9 @@ output:
-\BeginKnitrBlock{theorem}\iffalse{-91-65-32-84-104-101-111-114-101-109-93-}\fi{}
+\BeginKnitrBlock{theorem}{}

-\protect\hypertarget{thm:atheorem}{}{(\#thm:atheorem) \iffalse (A
-Theorem) \fi{} }This is a theorem. \EndKnitrBlock{theorem}
+ffalse\{-91-65-32-84-104-101-111-114-101-109-93-\}\fi{}
+\protect\hypertarget{thm:atheorem}{}{(\#thm:atheorem)
+\textbackslash iffalse (A Theorem) \fi{} }This is a theorem.
+\EndKnitrBlock{theorem}
````

This change of result is quite odd to me. It is like `\i` is not recognize as a raw LaTeX and get processed by Pandoc

This is the minimal dummy syntax I found that got transformed 

````powershell
❯ echo '\a{}\iffalse{}\fi{}' | pandoc -t latex
\a{}{}

ffalse\{\}\fi{}
````

I need to thing into this and looks into Pandoc changes to see if this is a bug on their side 🤔

**Comment by deleeuw** [comment]
I am using pandoc from homebrew

deleeuw@janspb:~/ > pandoc -v
pandoc 2.14.0.1
Compiled with pandoc-types 1.22, texmath 0.12.3, skylighting 0.10.5.1,
citeproc 0.4, ipynb 0.1.0.1
User data directory: /Users/deleeuw/.local/share/pandoc
Copyright (C) 2006-2021 John MacFarlane. Web:  https://pandoc.org
This is free software; see the source for copying conditions. There is no
warranty, not even for merchantability or fitness for a particular purpose.

**Comment by deleeuw** [comment]
The homebrew pandoc is now 2.14.0.2, but the problem persists. If I uninstall the homebrew pandoc (including pandoc-citeproc, pandoc-crossref and pandoc-include-code) and use the pandoc that comes with RStudio then the problem disappears. So its either pandoc or homebrew.

**Comment by yihui** [comment]
@cderv Thanks a lot for the careful investigation! It seems that we will have to deprecate the previous syntax sooner than we previously imagined. Alternatively and more practically, we can make the theorem/proof engines write out fenced Divs instead of using **knitr**'s block engine: https://github.com/rstudio/bookdown/blob/adfd645c75d0e8cef98e42d17557fadffb0d1452/R/utils.R#L464-L515 That means we would drop the support for very old versions of Pandoc that don't support fenced Div (which is okay to me), and also not support EPUB/Word (which is temporarily okay but we'd better improve it in the not so distant future).

**Comment by cderv** [comment]
That is indeed a nice solution instead (or before) deprecating. I think this would mean Pandoc 2.0+ as we already discussed several times (for Fenced div and Lua Filters)

However, I am wondering about EPUB. It seems quite used as a bookdown output and this would break output if they use those engines. 🤔 

Maybe do what you are suggestion but for newer Pandoc version to fix the current issue ? 
Or check the Pandoc version and drop support for this syntax and avise to use new one. Too much maybe 😅 

> but we'd better improve it in the not so distant future).

yes this needs to be scheduled and done.

**Comment by yihui** [comment]
> Maybe do what you are suggestion but for newer Pandoc version to fix the current issue ?

Yes, we can do that for newer Pandoc versions only, or use fenced Div for HTML and PDF output only. The latter way may make it easier for us to transition to the new syntax for all output formats.

**Comment by cderv** [comment]
I asked in pandoc discuss to understand if this change in Pandoc is expected.
https://groups.google.com/g/pandoc-discuss/c/nylIR9ZCSRg

**Comment by cderv** [comment]
No answer in Pandoc discuss, so I opened an issue: https://github.com/jgm/pandoc/issues/7460

**Comment by cderv** [comment]
Some notes: could it be linked to `latex_macros` extensions ? 🤔 
https://pandoc.org/MANUAL.html#extension-latex_macros

**Comment by yihui** [comment]
Should be fixed now. Thanks for the report!

You can test the development version via

```r
remotes::install_github('rstudio/bookdown')
```

**Comment by deleeuw** [comment]
Thanks, Yihui. 

—- Jan

**Comment by Scinawa** [comment]
Just for the record, if anyone has this bug in 2022, the solution could be the following: 

Go into the terminal of your RStudio

- write install.packages('remotes')
- write remotes::install_github('rstudio/bookdown')
- select the packages to update (or update all of them).

**Comment by cderv** [comment]
The fix for this issue has been included in **bookdown** 0.25 - so installing from CRAN should be enough to fix this issue. 

If not, and only the dev version  fixes it, then it means it is another issue that have recently been fixed but not directly related to this specific issue here.
